portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,31 +14,91 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+    //ArrayList<String> comments = new ArrayList<String>();
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    ArrayList<String> comments = new ArrayList<String>();
-    comments.add("This website is cool!");
-    comments.add("I like your pictures!");
-    comments.add("The colors look nice together!");
+    Query query = new Query("Comment").addSort("timestamp",SortDirection.DESCENDING);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+    
+    //equal string and check to see if it null
+    //if null set equal to default
+    String userNum = request.getParameter("user-comment-num");
+    int numberComments;
+    if (userNum == null){
+        numberComments = 5;
+    }else{
+        numberComments = Integer.parseInt(userNum);
+    }
+    int numberLoaded = 0;
+
+    ArrayList<Comment> comments = new ArrayList<Comment>();
+    for(Entity entity : results.asIterable()){
+        if (numberLoaded < numberComments){
+            long id = entity.getKey().getId();
+            String name = (String) entity.getProperty("name");
+            String text = (String) entity.getProperty("text");
+            long timestamp = (long) entity.getProperty("timestamp");
+
+            Comment comment = new Comment(id, name, text, timestamp);
+            comments.add(comment);
+            numberLoaded += 1;
+        }else{
+            break;
+        }
+    }
+    
+    String jsoncom = convertToJson(comments);
+
+    response.setContentType("application/json;");
+    response.getWriter().println(jsoncom);
+    }
+  
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException{
+    String name = request.getParameter("name-input");
+    String text = request.getParameter("text-input");
+    long timestamp = System.currentTimeMillis();
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    Entity commentEntity = new Entity("Comment");
+    commentEntity.setProperty("name",name);
+    commentEntity.setProperty("text",text);
+    commentEntity.setProperty("timestamp",timestamp);
 
-    String json = convertToJson(comments);
+    datastore.put(commentEntity);
+
+    response.setContentType("text/html;");
+    //response.getWriter().println(comments);
+    response.sendRedirect("/index.html");
+  }
 
-    response.setContentType("text/html;");//going to priint directly on page
-    response.getWriter().println(json);
+  private static String convertToJson(ArrayList<Comment> messages) {
+        Gson gson = new Gson();
+        return gson.toJson(messages);
     }
 
-  private String convertToJson(ArrayList<String> list){
+  private String convertToJsonW(ArrayList<String> list){
```

Comment:
Looks like this method can be removed now that you have integrated with Gson :).